src/mpc/cli/mpc_phase2_contribute.cpp

```diff
@@ -0,0 +1,158 @@
+#include "mpc_common.hpp"
+#include "snarks/groth16/mpc/phase2.hpp"
+
+using namespace libzeth;
+namespace po = boost::program_options;
+
+namespace
+{
+
+// Usage:
+//   $0 phase2-contribute [<options>] <challenge_file>
+//
+// Options:
+//   --out <file>        Response output file (mpc-response.bin)
+//   --digest <file>     Write contribution hash to file.
+//   --skip-user-input   Use only ststem randomness
+class mpc_phase2_contribute : public subcommand
+{
+private:
+    std::string challenge_file;
+    std::string out_file;
+    std::string digest;
+    bool skip_user_input;
+
+public:
+    mpc_phase2_contribute()
+        : subcommand("phase2-contribute")
+        , challenge_file()
+        , out_file()
+        , digest()
+        , skip_user_input(false)
+    {
+    }
+
+private:
+    void initialize_suboptions(
+        po::options_description &options,
+        po::options_description &all_options,
+        po::positional_options_description &pos) override
+    {
+        options.add_options()(
+            "out,o",
+            po::value<std::string>(),
+            "Reponse output file (mpc-response.bin)")(
+            "digest",
+            po::value<std::string>(),
+            "Write contribution digest to file")(
+            "skip-user-input", "Use only system randomness");
+        all_options.add(options).add_options()(
+            "challenge_file", po::value<std::string>(), "challenge file");
+        pos.add("challenge_file", 1);
+    }
+
+    void parse_suboptions(const po::variables_map &vm) override
+    {
+        if (0 == vm.count("challenge_file")) {
+            throw po::error("challenge_file not specified");
+        }
+        challenge_file = vm["challenge_file"].as<std::string>();
+        out_file = vm.count("out") ? vm["out"].as<std::string>()
+                                   : trusted_setup_file("mpc-response.bin");
+        digest = vm.count("digest") ? vm["digest"].as<std::string>() : "";
+        skip_user_input = (bool)vm.count("skip-user-input");
+    }
+
+    void subcommand_usage() override
+    {
+        std::cout << "Usage:\n  " << subcommand_name
+                  << " [<options>] <challenge_file>\n\n";
+    }
+
+    int execute_subcommand() override
+    {
+        if (verbose) {
+            std::cout << "challenge_file: " << challenge_file << "\n";
+            std::cout << "out_file: " << out_file << std::endl;
+            std::cout << "digest: " << digest << std::endl;
+            std::cout << "skip_user_input: " << skip_user_input << std::endl;
+        }
+
+        libff::enter_block("Load challenge file");
+        srs_mpc_phase2_challenge<ppT> challenge = [&]() {
+            std::ifstream in(
+                challenge_file, std::ios_base::binary | std::ios_base::in);
+            return srs_mpc_phase2_challenge<ppT>::read(in);
+        }();
```

Comment:
Why not use `read_from_file` that you defined in `mpc_common.hpp`?

Reply:
CHanged to use `read_from_file`